File: src/main/kotlin/nexters/weski/ski/resort/SkiResort.kt
package nexters.weski.ski.resort

import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import nexters.weski.common.BaseEntity
import nexters.weski.slope.Slope
import nexters.weski.webcam.Webcam
import java.time.LocalDate

@Entity
@Table(name = "ski_resorts")
class SkiResort(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val resortId: Long = 0,
    var name: String,
    @Enumerated(EnumType.STRING)
    var status: ResortStatus,
    var openingDate: LocalDate? = null,
    var closingDate: LocalDate? = null,
    var openSlopes: Int = 0,
    var totalSlopes: Int = 0,
    var dayOperatingHours: String? = null,
    var nightOperatingHours: String? = null,
    var lateNightOperatingHours: String? = null,
    var dawnOperatingHours: String? = null,
    var midnightOperatingHours: String? = null,
    var snowfallTime: String? = null,
    var xCoordinate: String,
    var yCoordinate: String,
    var xRealCoordinate: Double? = null,
    var yRealCoordinate: Double? = null,
    var detailedAreaCode: String,
    var broadAreaCode: String,
    @OneToMany(mappedBy = "skiResort", fetch = jakarta.persistence.FetchType.LAZY)
    val slopes: List<Slope> = emptyList(),
    @OneToMany(mappedBy = "skiResort", fetch = jakarta.persistence.FetchType.LAZY)
    val webcams: List<Webcam> = emptyList(),
) : BaseEntity() {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SkiResort) return false
        return resortId != 0L && resortId == other.resortId
    }

    override fun hashCode(): Int = 31

    override fun toString(): String = "SkiResort(resortId=$resortId, name='$name')"
}

enum class ResortStatus {
    운영중,
    운영종료,
    예정,
}

File: weski-admin/src/types/skiResort.ts
// API 응답 타입
export interface ApiResponse<T> {
  success: boolean
  message: string
  data?: T
}

// 스키장 상태 열거형
export type ResortStatus = '운영중' | '운영종료' | '예정'

// Admin용 스키장 응답 DTO
export interface AdminSkiResortResponse {
  resortId: number
  name: string
  status: string
  openingDate?: string
  closingDate?: string
  openSlopes: number
  totalSlopes: number
  dayOperatingHours?: string
  nightOperatingHours?: string
  lateNightOperatingHours?: string
  dawnOperatingHours?: string
  midnightOperatingHours?: string
  snowfallTime?: string
  xCoordinate: string
  yCoordinate: string
  detailedAreaCode: string
  broadAreaCode: string
  createdAt: string
  updatedAt: string
}

// 스키장 생성 요청 DTO
export interface CreateSkiResortRequest {
  name: string
  status: ResortStatus
  openingDate?: string
  closingDate?: string
  dayOperatingHours?: string
  nightOperatingHours?: string
  lateNightOperatingHours?: string
  dawnOperatingHours?: string
  midnightOperatingHours?: string
  snowfallTime?: string
  xCoordinate: string
  yCoordinate: string
  detailedAreaCode: string
  broadAreaCode: string
}

// 스키장 수정 요청 DTO
export interface UpdateSkiResortRequest {
  name?: string
  status?: ResortStatus
  openingDate?: string
  closingDate?: string
  dayOperatingHours?: string
  nightOperatingHours?: string
  lateNightOperatingHours?: string
  dawnOperatingHours?: string
  midnightOperatingHours?: string
  snowfallTime?: string
  xCoordinate?: string
  yCoordinate?: string
  detailedAreaCode?: string
  broadAreaCode?: string
}

// 슬로프 타입
export interface Slope {
  slopeId: number
  name: string
  difficulty: string
  isDayOperating: boolean
  isNightOperating: boolean
  isLateNightOperating: boolean
  isDawnOperating: boolean
  isMidnightOperating: boolean
  webcamNo?: number
}

// 슬로프 수정 요청 DTO
export interface UpdateSlopeRequest {
  name: string
  difficulty: string
  webcamNumber?: number
  isDayOperating: boolean
  isNightOperating: boolean
  isLateNightOperating: boolean
  isDawnOperating: boolean
  isMidnightOperating: boolean
}

// 웹캠 타입
export interface Webcam {
  id: number
  name: string
  number: number
  description?: string
  url?: string
  isExternal?: boolean
}

// 웹캠 수정 요청 DTO
export interface UpdateWebcamRequest {
  name?: string
  description?: string
  url?: string
}
